File: src/main/java/com/dude/dms/backend/data/DataEntity.java
package com.dude.dms.backend.data;

import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.MappedSuperclass;
import java.io.Serializable;
import java.util.Objects;

@MappedSuperclass
public abstract class DataEntity implements Serializable {

    @Id
    @GeneratedValue
    private Long id;

    public Long getId() {
        return id;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof DataEntity) {
            return ((DataEntity) obj).getId().equals(id);
        }
        return false;
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
}

File: src/main/kotlin/com/dude/dms/backend/data/DataEntity.kt
package com.dude.dms.backend.data

import java.io.Serializable
import java.util.*
import javax.persistence.GeneratedValue
import javax.persistence.Id
import javax.persistence.MappedSuperclass

@MappedSuperclass
abstract class DataEntity : Serializable {

    @Id
    @GeneratedValue var id = 0L

    override fun equals(other: Any?) = if (other is DataEntity) other.id == id else false

    override fun hashCode() = Objects.hash(id)
}